stock_reserve_sale/model/sale.py

```diff
@@ -0,0 +1,160 @@
+        for sale in self:
+            has_stock_reservation = False
+            is_stock_reservable = False
+            for line in sale.order_line:
+                if line.reservation_ids:
+                    has_stock_reservation = True
+                if line.is_stock_reservable:
+                    is_stock_reservable = True
+            if sale.state not in ("draft", "sent"):
+                is_stock_reservable = False
+            sale.is_stock_reservable = is_stock_reservable
+            sale.has_stock_reservation = has_stock_reservation
+
+    has_stock_reservation = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
+        store=True,
+        string="Has Stock Reservations",
+    )
+    is_stock_reservable = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
+        store=True,
+        string="Can Have Stock Reservations",
+    )
+    reserves_count = fields.Integer(compute="_compute_reserves_count")
+    all_lines_reserved = fields.Boolean(
+        compute="_compute_reserves_count", store=True, default=False
+    )
+
+    def release_all_stock_reservation(self):
+        line_ids = [line.id for order in self for line in order.order_line]
+        lines = self.order_line.browse(line_ids)
+        lines.release_stock_reservation()
+        return True
+
+    def _compute_reserves_count(self):
+        reserve_ids = self.env["stock.reservation"]._read_group(
+            domain=expression.AND(
+                [
+                    [("sale_id", "in", self.ids)],
+                ]
+            ),
+            fields=["sale_id"],
+            groupby=["sale_id"],
+        )
+        lines = self.order_line.filtered(
+            lambda l: l.product_id.detailed_type != "service"
+        )
+        for order in self:
+            if reserve_ids:
+                order.reserves_count = reserve_ids[0]["sale_id_count"]
+            else:
+                order.reserves_count = 0
+            if order.reserves_count == len(lines):
+                order.all_lines_reserved = True
+            else:
+                order.all_lines_reserved = False
+
+    def action_view_reserves_products(self):
+        action = self.env["ir.actions.act_window"]._for_xml_id(
+            "stock_reserve.action_stock_reservation_tree"
+        )
+        action["domain"] = [("sale_id", "in", self.ids)]
+        action["context"] = {"search_default_groupby_product": False}
+        return action
+
+    def action_view_reserves_stock_picking(self):
+        stock_picking = ""
+        action = self.env["ir.actions.actions"]._for_xml_id(
+            "stock.action_picking_tree_all"
+        )
+        stock_picking = (
+            self.env["stock.picking"]
+            .search([("origin", "=", self.name)])
+            .filtered(lambda a: a.state not in "cancel")
+        )
```

Comment:
I think we can make a search with both conditions.
```suggestion
        stock_picking = (
            self.env["stock.picking"]
            .search([
                ("origin", "=", self.name),
                ("state", "!=", "cancel")
            ])
        )
```

Reply:
Generally a != condition for a state field in search is a bad idea. The problem with these is often state fields are indexed and the other conditions aren't. But because of the generally bad distribution of states it is usually faster if you aren't using '=' for a state which is a small percentage, the queries can be quite bad. The search would have prefetched anyway, the cancelled records will be minimal. IDK, I feel like I would want testing on big databases first as one of the performance bottlenecks we always look for is exactly this suggestion. It is almost always better to reverse it to a state in type query or use filtered.

Reply:
Just looking at this, the filtered clause is wrong anyway, it should be != instead of not in. But as it happens, in testing the suggested query on a database with 1m pickings performed basically identically as the original BEFORE the filtered clause and planned identically using standard Odoo indexes, so I think it is a good suggestion.

Reply:
@gdgellatly Thanks for your tests

Reply:
@gdgellatly Have you tried with a condition like "state", "in", "[....]" ? Did they give the same result (in time)?

Reply:
Doesn't really matter I think. The existing code is clearly a mistake. But also the planner will have statistics on uniqueness and origin field is indexed so regardless, for this query, I think no matter the state condition, it will pick the origin index.